### ISSUE TITLE
Features in environment

### ISSUE THREAD

**Issue by fabianvdW** [rust-lang/cargo#8157]
**Problem**

Suppose I have a workspace, containing crate `A`,`B` and `C`.  `A` has the feature `F`, but it is not a default-feature. Both `B` and `C` depend on `A`, however `B` does not use `F`, while `C` does.

The file `./A/src/lib.rs` looks like this:
```Rust
pub fn A(){
    if cfg!(feature = "F"){
        println!("Using library A with Feature F")
    }else {
        println!("Using library A")
    }
}
```
, while B and C just call `A::A()` in their main function.
Behaviour:
```
cargo run --bin B
Using library A with Feature F

cargo build
"./target/debug/B.exe"
Using library A with Feature F

cargo run -p B
Using library A
```
This behaviour just seems inconsistent. I don't understand how B is ever built with F, as in my real use case this imposes a performance penalty for B, so I will have to fall back to using `cargo run -p B`. The behaviour I expect is:
```
cargo run --bin B
Using library A

cargo build
"./target/debug/B.exe"
Using library A

cargo run -p B
Using library A
```
**Steps**

1. Make a workspace containing three crates A , B, C.
2. 
`A/src/lib.rs` as above
`B/src/main.rs`: 
```Rust
fn main() {
    A::A();
}
```
`A/Cargo.toml`:
```
[features]
default = []
F = []
```
`B/Cargo.toml`
```
[dependencies]
A = {path = "../A"}
```
`C/Cargo.toml`
```
[dependencies.A]
path = "../A"
default-features = false
features = ["F"]
```
3. Run cargo commands as above

**Possible Solution(s)**

**Notes**

Output of `cargo version`:
`cargo 1.43.0 (3532cf738 2020-03-17)`
Using latest stable Rust, `rustc --version`:
`rustc 1.43.0 (4fb7144ed 2020-04-20)`

**Comment by ehuss** [comment]
I can see how that can be confusing.  `cargo run --bin B` has an implicit `--workspace` flag which means it is building all workspace members, and then filtering out the binary `B`.  Workspace features are unified for all selected packages.  Feature resolution looks at which packages are being built, but doesn't know about the filtering operation (which is done after resolution).

It may be very difficult to work around that. Generating the [target proposals](https://github.com/rust-lang/cargo/blob/80cf934ae63b3727348b666a3d3af37e39587bc1/src/cargo/ops/cargo_compile.rs#L781-L782) requires having resolution already being performed.

It may be possible to special-case this scenario and constrain the initial specs (so it would infer if you did `--bin B` that you also implied `-p B`), but I can imagine a lot of edge cases that would make that tricky to get right.

**Comment by fabianvdW** [comment]
But is it actually wanted that even when all packages are built, `A` isn't build twice, once with `F` and without `F`? Atleast in my use-case this is what I would expect from `cargo build`.

**Comment by ehuss** [comment]
It's questionable, some projects will want the unification (because it cuts down on compile time), and some do not (because it can alter behavior or prevent compilation).  This is tracked in #4463 among others.

**Comment by fabianvdW** [comment]
I see, in that this case fixing this will just be a minor enhancement, as I can just use the working command. It would be nice if one could just configure such things for each workspace.

**Comment by Veetaha** [comment]
@ehuss 
> cargo run --bin B has an implicit --workspace flag which means it is building all workspace

I don't think so, apparently `carg run --bin` compiles only what is needed to build the binary (this is very noticable in my project which is a huge cargo workspace) which is what `--bin` flags' purpose is, isn't it?

**Comment by ehuss** [comment]
As explained above, it selects the entire workspace, and then filters it down to whichever package contains the named binary. The sequence is roughly:

1. Select entire workspace (implied `--workspace`).
2. Resolve features across entire workspace.
3. Filter selected packages to find one with the named binary.
4. Build that binary.

It may be possible to move step 3 above step 2 (so it resolves features only for the selected package), but I think it will be tricky.

**Comment by epage** [comment]
FYI I've posted rust-lang/rfcs#3692

**Comment by ehuss** [comment]
The cargo team discussed this ([notes](https://github.com/rust-lang/cargo-team/blob/HEAD/meetings/sync-meeting/2026-06-09.md#backlog---bin-b-resolves-features-differently-than--p-b-in-a-workspace)). I think we were generally +1 to fixing this so that `--bin foo` would not have an implicit `--workspace`. However, we were uncertain on how complex that would be to implement. If the implementation isn't too complicated, then I think it should be fine to change.

In terms of implementation, I'm wondering if it would be possible for the early CLI parsing code to detect this situation, look at the workspace and try to find the packages with the given binary name, and then add implicit `-p` flags for those workspace packages.

**Comment by epage** [comment]
> In terms of implementation, I'm wondering if it would be possible for the early CLI parsing code to detect this situation, look at the workspace and try to find the packages with the given binary name, and then add implicit -p flags for those workspace packages.

This sounds very similar to the idea of "when a `--bin` is specified, automatically activate `required-features`, so there is the potential of multiple benefits from any complexity involved here.
